feat: update observation photo preview screen

achou11 opened the pull request:
Towards #1043

I'm expecting there to be a few more changes needed in this PR before it can be merged, but clarification is needed in a few places.

- Most of the metadata mentioned in the issue is not exposed because we need to upgrade core to account for those newly added fields. That will be done in a separate PR. For now, this PR is mostly focused around the UI changes for the relevant screens, and we render as much of the necessary info as possible for the time being.

- The designs didn't seem to indicate any changes to how deletion of a draft photo works. I've kept it the same as what currently exists (button in the top-right of the screen). The major difference is the implementation of the confirmation flow, which is now a separate navigation screen, which aligns with other similar flows.

- If you're testing the deletion of draft photos, you may run into #1172. That's being fixed in #1174.

- Still not sure what "validated by CoMapeo" means or the proper way to calculate that. Appreciate any insights.

---

### Observation create view

- View draft photo

    <img width="300" alt="observation-create_processed-draft-photo" src="https://github.com/user-attachments/assets/e596590e-2f61-48ef-91eb-fc17b296beff" />

- Draft photo deletion interaction 

    <img width="300" alt="observation-create_processed-draft-photo-delete-confirmation" src="https://github.com/user-attachments/assets/6cab8fdc-3271-451c-b5fc-59d04c801bbc" />

    https://github.com/user-attachments/assets/05c62871-6c80-405a-ae95-33038ea49e5a
    
### Saved observation view

- Viewing saved observation photo 

    <img width="300" alt="view-observation_edit-observation_saved-photo" src="https://github.com/user-attachments/assets/fb1a367d-efc8-46b9-be26-c7cbbf9ee235" />

### Edit observation view

- View draft photo

    <img width="300" alt="observation-edit_processed-draft-photo" src="https://github.com/user-attachments/assets/7cfe1bd6-6cb6-489a-8932-cb8f0eea9b25" />

### Error placeholders

- Draft photo
    
    <img width="300" alt="error-placeholder_processed-draft-photo" src="https://github.com/user-attachments/assets/60747448-c6e5-4fcf-b711-95fcf471f631" />

- Saved photo

    <img width="300" alt="error-placeholder_saved-photo" src="https://github.com/user-attachments/assets/b6a33b40-9991-40f0-a703-56c053e4954b" />

achou11 committed:
initial implementation

achou11 committed:
image component abstractions

achou11 commented on src/frontend/sharedTypes/navigation.ts:
```diff
@@ -43,7 +43,16 @@ export type RootStackParamsList = {
+    validatedByCoMapeo?: boolean;
+  };
+  ConfirmDeletePhoto: {
+    onSuccess?: () => void;
```
doing this aligns with our agreed pattern of allowing callback nav params for certain kinds of screens.

In a separate branch, I experimented with a non-callback based approach that seems to work, and aligns with some of the suggestions that React Navigation provides to avoid callbacks as nav params.

https://github.com/digidem/comapeo-mobile/commit/a4eb915d6914716108caa43cb7aa620c40fae6ca

Not sure if it's really that much better, but it works and isn't overly complicated.

cimigree replied:
I like the way it is in here better if that is what you are asking.

achou11 commented on src/frontend/sharedTypes/navigation.ts:
```diff
@@ -43,7 +43,16 @@ export type RootStackParamsList = {
   CoordinateFormat: undefined;
   Experiments: undefined;
   PhotoPreviewModal: {
+    createdByDeviceId?: string;
```
I opted to make this a nav param as opposed to something that this screen fetches internally, with the intention that this screen does as little work as possible when it comes to fetching data.

Open to changing but I think it's generally worked out okay so far 😄

cimigree replied:
Will all 5+ (not sure how photo metadata will be achieved/ received/ broken down) properties be passed in here also for the same reason? That seems like a lot...

achou11 replied:
I suspect that the photo metadata will live within the `photo` nav prop, so no additional nav params should be needed to access those when they're introduced.

achou11 commented on src/frontend/sharedComponents/Images/TrulyContainedImage.tsx:
this was painful to figure out but pretty happy with the result! React Native's behavior for `Image`s with `resizeMode="contain"` is pretty rigid, and doesn't account for dynamically sized containers/images very well. In most cases, it will add padding to fill-in the spaces that are unoccupied in the container due to the resizing of the image, which is undesirable in our case.

<img width="300" src="https://github.com/user-attachments/assets/f7f7645f-a4a4-4612-86fe-fc651996f2c6" />

achou11 commented on src/frontend/sharedComponents/Images/GracefulImage.tsx:
Got annoyed with the amount of repetition that was needed to elegantly handle image errors. Decided to create this abstraction to make it easier to implement.

Open to name changes. This was unfortunately the best I could come up with...

cimigree replied:
Frankly I find the wrappers around wrappers around wrappers a little confusing. I assume you had to do it this way? If there is a way to pull at least one layer back on the wrappers, I think that would be preferable, more readable, easier to follow... Not sure about testing though. Frankly I am not really sure what this component is doing so I can't really hazard a suggestion as to naming.

achou11 replied:
Yeah I think we can afford to get rid of this abstraction if preferable. It was mainly introduced to introduce a consistent way of displaying images that may fail to load (e.g. anything that relies on a network).

Without this, every image that could potentially fail to load will need similar boilerplate of adding an `onError()` handler for the image, and then a stateful variable to indicate that an error occurred and something else should be rendered.

achou11 commented on src/frontend/sharedComponents/Editor/index.tsx:
```diff
@@ -131,7 +134,45 @@ export const Editor = ({
+                            return;
+                          }
+
+                          // TODO: Ideally use the core-react hooks but the error handling makes it tricky
```
didn't want to go down a rabbithole of making changes to something like the Editor component to make this more elegant. in the context of what this is for, I think this is generally okay for now.

alternatively, this could be removed if we decide to make the PhotoPreviewModal fetch the `createdByDeviceId` value internally as opposed to supplying it via a nav param. Wasn't too keen on doing that as an initial implementation, but open to considering it if necessary.

cimigree replied:
All of this is happening in onPress? Seems difficult to read and a little fragile. You are only catching errors at one level. Is this the best way?

achou11 replied:
the top level `catch` should also handle catches from the internal promise (note the returning of the internal promise i.e. `return projectApi.$originalVersionIdToDeviceId(...`).

whether this is the best way is debatable! like the comment mentions, i think it's partially a symptom of how this component is set up, but maybe moving the fetching to the photo preview screen is what we should do anyways.

achou11 replied:
ended up moving the loading of this info into the screen. see https://github.com/digidem/comapeo-mobile/pull/1179/commits/f7961cf39c5b0eafb3b4c2d5457efa244241d0c4

achou11 commented on src/frontend/sharedComponents/Editor/index.tsx:
```diff
@@ -131,7 +134,45 @@ export const Editor = ({
+                              });
+                            })
+                            .catch(() => {
+                              // TODO: Does it make sense to navigate to the screen if we fail to retrieve `createdByDeviceId`?
+                              navigate('PhotoPreviewModal', {
+                                photo: att,
+                                observationDocId: observationId,
+                                validatedByCoMapeo: true,
+                              });
```
Highlighting this. not sure how often this would happen and I figured navigating to the screen with some missing info is better than not doing so at all.

separately, should this log to sentry if this happens?

achou11 commented on src/frontend/sharedTypes/navigation.ts:
```diff
@@ -43,7 +43,16 @@ export type RootStackParamsList = {
+    createdByDeviceId?: string;
+    observationDocId?: string;
     photo: SavedPhoto | ProcessedDraftPhoto;
+    validatedByCoMapeo?: boolean;
```
depending on how this is determined, may not need to be a nav param. but my lack of certainty has led me to use this as such for now.

cimigree replied:
The validated by Comapeo? In Observation Metadata it is a given if the location is known and determined during taking the observation, not when it is added by the user. Maybe that will be the same thing here?

achou11 replied:
Ah just found the note in notion that provides some insight on this:

> Validated by CoMapeo (data is data that has been collected automatically and not edited)  accompanied by the Check Dot icon

honestly, still not quite sure what that means on a technical level. Will ask for clarification and post here

achou11 replied:
From Aldo:

> To my understanding any photo taken using CoMapeo is considered Validated.
> In the instance (I think in the future because this is not possible today) that a photo is uploaded manually than we can't consider it Validated. We will instead show "Uploaded Manually".

Which makes more sense, but as far as I know, there's currently no definitively way to derive this information from a technical level (e.g. a metadata field like `takenByCoMapeo`).

> In Observation Metadata it is a given if the location is known and determined during taking the observation, not when it is added by the user. Maybe that will be the same thing here?

I think it's different in this case because it's referring to the validity of the photo, not the observation. As mentioned above, there can be a future where some of the photos associated with an observation may not have been created via a CoMapeo application, which is the thing that this UI is trying to highlight I guess.

achou11 replied:
based on discussions, seems like we're going to go with checking that the photo has a timestamp and location as part of the metadata for now. this definitely has its limitations (that I've raised concerns about) but we don't really have the pieces in place to do this check more properly.

achou11 replied:
implemented ^ via https://github.com/digidem/comapeo-mobile/pull/1179/commits/713efd12e2b687377ccca4b4c23e48f460a4d766

achou11 commented:
Just realized that one thing I haven't tested yet is viewing synced photos, especially from the device that receives them. Most notably should test that if the media sync settings are set to previews only, that rendering the photo in the preview screen works as expected.

achou11 commented on src/frontend/sharedComponents/Images/CoreBlobImage.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  const {
+    data: attachmentUrl,
+    error,
```
does using this error actually work? i'm under the impression that in the case of the initial query failing, the use of a suspense hook here is actually going to propagate to the nearest error boundary (for which we have none!!!), meaning that this component won't be rendered.

achou11 replied:
~just tested this by setting the "variants" to fetch to some nonsense and it works as intended! kind of surprised because i thought react query would've attempted to trigger an error boundary 🤔~ EDIT: nevermind, changing the values at the hook level isn't testing what I'm asking about. to properly test, I need to simulate core throwing an error when trying to build the attachment url

cimigree replied:
Ok I don't think I can do that.

achou11 replied:
came to the conclusion that attempting to use the `error` here doesn't work as intended in the case of the initial query resulting in an error. Ended up moving the error handling to the consuming component and using error boundaries. see https://github.com/digidem/comapeo-mobile/pull/1179/commits/a660e64bd62f50a8318adf439093cd694d2e61ef

for future reference, a quick way to force test suspense errors is to implement a bad query in the component. for example, I added this to the component that uses the core-react hooks:

```ts
const badQuery = useSuspenseQuery({ queryKey: ['bad'], queryFn: () => throw new Error('Failed') })
```

This allowed me to test how the error boundaries I added are used.

cimigree commented on src/frontend/screens/PhotoPreviewModal.tsx:
```diff
@@ -1,133 +1,174 @@
+        )}
+      </View>
+      <View style={{gap: 20}}>
+        {validatedByCoMapeo && (
```
On my two devices, the photo looks good but the data properties don't quite align with the icons. Can you fix that?
<image src="https://github.com/user-attachments/assets/ffdd1d52-0d4a-4cb3-bdf1-7a5a336212fb" width="250" />

achou11 replied:
addressed via https://github.com/digidem/comapeo-mobile/pull/1179/commits/13823bd9eae2ab7533d85ce1483dd883822cd73c

cimigree replied:
FYI, They align with regular size text and display but not with enlarged display.
<image src="https://github.com/user-attachments/assets/625c6b30-4ca5-4cb8-ac1a-445a5538ac8e" width="250" />

achou11 replied:
interesting - seems to vary between devices...i'm not noticing as much of a misalignment (if any) on my Pixel devices:

- Pixel 6, with max font size:

  <img width="300" alt="image" src="https://github.com/user-attachments/assets/5a6ae251-5c42-4e4e-b31b-c23622345ba6" />
  
- Pixel 2, with max font size:

    <img width="300" alt="image" src="https://github.com/user-attachments/assets/c1cf5128-00d5-47b2-a549-914596f5d55f" />

But with my Samsung, it's definitely looking similar to yours in terms of misalignment:

- Samsung, with max font size:

    <img width="300" alt="image" src="https://github.com/user-attachments/assets/d99fef6a-1c7b-4ca7-8c4d-a55351cc06b8" />

---

Not entirely sure what a good fix is. Could do some math to adjust the padding of the icons based on the font multiplier I guess...

achou11 replied:
Think I figured it out! See https://github.com/digidem/comapeo-mobile/pull/1179/commits/edc77f16a57376279874bfd13f6f86cdf181d252

- Pixel 6 (max text size):

    <img width="300" alt="image" src="https://github.com/user-attachments/assets/da391815-9228-4f8a-aae5-c7a797ee788c" />

- Pixel 2 (max text size):

    <img width="300" alt="image" src="https://github.com/user-attachments/assets/d491a881-4642-427e-911b-2ea62416b639" />

- Samsung (max text size):

    <img width="300" alt="image" src="https://github.com/user-attachments/assets/6385a101-b0f3-43ea-b44a-b69495b83133" />

cimigree reviewed (approved):
I can't seem to find the sync settings (preview, etc that you asked about). I think they got accidentally removed from the new menu? Oops?

Are landscaped photos supposed to work? I never tried before so not sure. See the attached photo with icon misalignment.

I really don't know what to say about this. I find the logic confusing and the wrappers difficult to trace and understand, but perhaps that is the best architecture? I don't honestly know. Perhaps it is fine for now and when we get the metadata, it could be refactored? I feel like you know a lot more about how the app works and maybe it has to be this complicated.

achou11 commented:
> I can't seem to find the sync settings (preview, etc that you asked about). I think they got accidentally removed from the new menu? Oops?

Yeah it was removed, I temporarily added it back locally to test what I was interested in 😄 

> Are landscaped photos supposed to work? I never tried before so not sure. See the attached photo with icon misalignment.

Based on the designs, yes! However, I think the fact that it doesn't show in landscape in this preview is more related to how the photo is taken because we don't allow configuring the rotation before saving the image. In other words, I think how it gets displayed in this screen is outside the scope of this PR because it's information that's encoded when you take the photo.

> I really don't know what to say about this. I find the logic confusing and the wrappers difficult to trace and understand, but perhaps that is the best architecture? I don't honestly know. Perhaps it is fine for now and when we get the metadata, it could be refactored? I feel like you know a lot more about how the app works and maybe it has to be this complicated.

I think this is a fair criticism! Can look into removing some of the abstractions (e.g. the graceful image component), if that's what you're referring to.

achou11 committed:
Merge branch 'develop' into 1043/update-photo-preview-ui

achou11 commented:
> Based on the designs, yes! However, I think the fact that it doesn't show in landscape in this preview is more related to how the photo is taken because we don't allow configuring the rotation before saving the image. In other words, I think how it gets displayed in this screen is outside the scope of this PR because it's information that's encoded when you take the photo.

To follow-up on this, I guess we could try to be clever and decide whether or not to rotate the image based on something like the dimensions, but I think that's kind of fragile (how do you know which way to rotate??). Would rather we fix the source of the issue, which I believe is related to the photo-taking implementation.

achou11 committed:
fix vertical alignment of icon with text in info items

achou11 committed:
move loading of created by device id into PhotoPreviewModal

achou11 committed:
update how validation is calculated

achou11 committed:
remove GracefulImage abstraction

achou11 committed:
insert error boundaries

achou11 commented on src/frontend/screens/ConfirmDeletePhoto.tsx:
```diff
@@ -0,0 +1,92 @@
+  },
+  deleteImage: {
+    id: 'screens.ConfirmDeletePhoto.deleteImage',
+    defaultMessage: 'Delete Image',
```
Should I use `Photo` instead of `Image` here? This is from the existing language but feels slightly inconsistent with the wording we're using elsewhere

cimigree replied:
Yes I think so! Photo makes more sense to me to be consistent.

achou11 replied:
addressed via https://github.com/digidem/comapeo-mobile/pull/1179/commits/afeb6f13b9e9fe2b6fd74632e4d10f19fbf4ce4b

achou11 commented:
@cimigree think I addressed most of the feedback! Regarding some of the component abstractions, decided to remove the `GracefulImage` one but kept the rest

cimigree commented:
> @cimigree think I addressed most of the feedback! Regarding some of the component abstractions, decided to remove the `GracefulImage` one but kept the rest

Thanks that is helpful to drill down to only one to two additional layer(s) of abstraction.
Will be eager to see this with all of the metadata.

achou11 committed:
update translation in ConfirmDeletePhoto

achou11 committed:
fix icon size when OS font size is adjusted
